app/features/ai_resistant_assignment_generator/tools.py

```diff
@@ -0,0 +1,172 @@
+        self.model = ChatGroq(model=model_name,temperature=0.3,api_key="")
+
+    def read_text_file(self,filepath):
+
+        with open(f"{self.path}{filepath}", 'r') as file:
+            return file.read()
+
+    def build_prompt(self,filepath):
+        # build invididual promps for each sub part of syllabus
+        template = self.read_text_file(filepath)
+
+        prompt = PromptTemplate.from_template(template)
+        return prompt
+
+    def validator(self,response):
+        data = []
+        try:
+            data = json.loads(response)
+        except json.JSONDecodeError as e:
+
+           
+            print("JSON Decode Error , Trying to correct the JSON")
+            try:
+                corrected_result = response[min(response.find('{'),response.find('[')):max(response.rfind(']'),response.rfind('}')) + 1]
+                data = json.loads(corrected_result)
+                print("Corrected Parsed JSON successfully")
+            except json.JSONDecodeError as e:
+                print("Failed to parse corrected JSON")
+        return data
+    
+    def extract_text_from_txt(self,file_content):
+        return file_content.decode('utf-8')
+
+    def extract_text_from_pdf(self,file_content):
+        pdf_reader = PdfReader(BytesIO(file_content))
+        text = ''
+        for page in pdf_reader.pages:
+            text += page.extract_text()
+        return text
+
+    def extract_text_from_docx(self,file_content):
+        doc = Document(BytesIO(file_content))
+        text = '\n'.join([para.text for para in doc.paragraphs])
+        return text
+
+    def download_file(self,url):
+        response = requests.get(url)
+        if response.status_code == 200:
+            content_type = response.headers['Content-Type']
+            return response.content, content_type
+        else:
+            raise Exception(f"Failed to download the file. Status code: {response.status_code}")
+
+    def extract_content_from_url(self,file_url):
+        file_content, content_type = self.download_file(file_url)
+
+        # Check the Content-Type header for file type
+        if 'application/pdf' in content_type:
+            print("Extracting content from PDF...")
+            return self.extract_text_from_pdf(file_content)
+        elif 'application/vnd.openxmlformats-officedocument.wordprocessingml.document' in content_type:
+            print("Extracting content from Word Document (.docx)...")
+            return self.extract_text_from_docx(file_content)
+        elif 'text/plain' in content_type:
+            print("Extracting content from Text File...")
+            return self.extract_text_from_txt(file_content)
+        else:
+            raise Exception(f"Unsupported file type: {content_type}. Unable to extract content.")
+
+    def run(self):
+        # Important study resources and their specific function
+
+        prompt = self.build_prompt('ABSOLUTE FILE PATH')
+        chain = prompt | self.model
```

Comment:
You can add a `JSONOutputParser` to your LCEL chain for having deterministic outputs.